tls/src/params.cc

```diff
@@ -216,7 +228,8 @@ void params::validate_cert(gnutls_session_t session) {
   if (!_ca.empty()) {
     int ret;
     uint32_t status;
-    ret = gnutls_certificate_verify_peers2(session, &status);
+    ret = gnutls_certificate_verify_peers3(session, _tls_hostname.c_str(),
+                                           &status);
```

Comment:
Are you sure this will work?
In case of a given tls_hostname, I have no doubt. But in case of an empty tls_hostname, I think gnutls_certificate_verify_peers3 needs a nullptr value instead of an empty string. Am I wrong?

So, two cases here:
* The string is not empty and your code is good.
* The user gave nothing, so here you enter an empty string and I think the check will happen badly.

Correct me if I'm wrong.

Reply:
This comes from the documentation here: https://man7.org/linux/man-pages/man3/gnutls_certificate_verify_peers3.3.html


If the  hostname provided is non-NULL then this function will
       compare the hostname in the certificate against it. The
       comparison will follow the RFC6125 recommendations. If names do
       not match the GNUTLS_CERT_UNEXPECTED_OWNER status flag will be
       set.